### PR TITLE
Fix infinite loop in tagging

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1927519

Introduced with https://github.com/dotnet/roslyn/pull/70884.  I accidentally moved an index that was outside of a loop into the loop.

### CODE REVIEW

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs [hunk 1]
```diff
@@ -134,49 +139,62 @@ private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection reque
...
 
             if (!enumerator.MoveNext())
                 return;
 
             using var _2 = PooledHashSet<ITagSpan<TTag>>.GetInstance(out var hashSet);
 
+            var requestIndex = 0;
             while (true)
             {
-                var requestIndex = 0;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:150]: this was hte mistake when i did the original refactoring here.  by resetting requestIndex to 0 at hte top of hte loop we were effectively never moving that  item forward in the second if-check.

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs [hunk 2]
```diff
@@ -134,49 +139,62 @@ private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection reque
...
                 var currentTag = enumerator.Current;
 
                 var currentRequestSpan = requestedSpans[requestIndex];
                 var currentTagSpan = currentTag.Span;
 
-                if (currentRequestSpan.Start > currentTagSpan.End)
+                // The current tag is *before* the current span we're trying to intersect with.  Move to the next tag to
+                // see if it intersects with the current span.
+                if (currentTagSpan.End < currentRequestSpan.Start)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:165]: swapped the direction as i felt it was easier to understand.

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs [hunk 3]
```diff
@@ -134,49 +139,62 @@ private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection reque
...
                 {
+                    // If there are no more tags, then we're done.
                     if (!enumerator.MoveNext())
-                        break;
+                        return;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:169]: thought it was clearer to indicate the entire algorithm is done, as opposed to breaking the loop, which then exits the method.

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs [hunk 4]
```diff
@@ -53,11 +53,16 @@ public bool HasSpanThatContains(SnapshotPoint point)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:53]: view with whitespace off.

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs [hunk 5]
```diff
@@ -134,49 +139,62 @@ private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection reque
...
+
+                    continue;
                 }
-                else if (currentTagSpan.Start > currentRequestSpan.End)
+
+                // The current tag is *after* teh current span we're trying to intersect with.  Move to the next span to
+                // see if it intersects with the current tag.
+                if (currentTagSpan.Start > currentRequestSpan.End)
                 {
                     requestIndex++;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:178]: this is the condition that was not working anymore.  TO hit the bug you need:

1. a request for > 100 spans.
2. the first span requested has to be before *any* of the tags and not intersect it.

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:150]: You can see here that this was appropriately outside the loop prior to the mistake: https://github.com/dotnet/roslyn/pull/70884/files?diff=split#diff-9d0243eeadbc7ee94e5712ce574223cb596fa3a6110785a8e34e9e90dcc98e55L134

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @genlu ptal

**Review by arkalyanms — APPROVED** [review]

**Pull request merged**